log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/util/PluginManagerTest.java

```diff
@@ -0,0 +1,224 @@
+import org.apache.logging.log4j.core.config.Configuration;
+import org.apache.logging.log4j.core.config.ConfigurationSource;
+import org.apache.logging.log4j.core.config.Node;
+import org.apache.logging.log4j.core.config.plugins.processor.PluginCache;
+import org.apache.logging.log4j.core.config.plugins.processor.PluginEntry;
+import org.apache.logging.log4j.core.config.xml.XmlConfigurationFactory;
+import org.apache.logging.log4j.core.test.Compiler;
+import org.apache.logging.log4j.status.StatusData;
+import org.apache.logging.log4j.test.ListStatusListener;
+import org.apache.logging.log4j.test.junit.UsingStatusListener;
+import org.apache.logging.log4j.util.LoaderUtil;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class PluginManagerTest {
+
+    private static final String PLUGIN_DESCRIPTOR =
+            "META-INF/org/apache/logging/log4j/core/config/plugins/Log4j2Plugins.dat";
+
+    @BeforeEach
+    void setUp() {
+        PluginRegistry.getInstance().clear();
+    }
+
+    @Test
+    @UsingStatusListener
+    @SuppressWarnings("deprecation")
+    void calling_addPackage_issues_warning(final ListStatusListener listener) {
+        try {
+            PluginManager.addPackage("com.example");
+            assertThat(listener.findStatusData(Level.WARN))
+                    .anySatisfy(PluginManagerTest::assertContainsPackageScanningLink);
+            listener.clear();
+
+            PluginManager.addPackages(singletonList("com.example"));
+            assertThat(listener.findStatusData(Level.WARN))
+                    .anySatisfy(PluginManagerTest::assertContainsPackageScanningLink);
+            listener.clear();
+        } finally {
+            PluginManager.clearPackages();
+        }
+    }
+
+    @Test
+    @UsingStatusListener
+    void using_packages_issues_warning(final ListStatusListener listener) {
+        final ConfigurationSource source = ConfigurationSource.fromResource(
+                "customplugin/log4j2-741.xml", getClass().getClassLoader());
+        assertThat(source).as("Configuration source").isNotNull();
+        final Configuration configuration = new XmlConfigurationFactory().getConfiguration(null, source);
+        assertThat(configuration).as("Configuration").isNotNull();
+        configuration.initialize();
+        assertThat(configuration.getPluginPackages()).as("Plugin packages").contains("customplugin");
+        assertThat(listener.findStatusData(Level.WARN))
+                .anySatisfy(PluginManagerTest::assertContainsPackageScanningLink);
+    }
+
+    @Test
+    @UsingStatusListener
+    void using_packages_finds_custom_plugin(final ListStatusListener listener) throws Exception {
+        // To ensure our custom plugin is NOT included in the log4j plugin metadata file,
+        // we make sure the class does not exist until after the build is finished.
+        // So we don't create the custom plugin class until this test is run.
```

Comment:
Would you mind also documenting `FixedStringLayout.java.source` with similar 3-4 sentences to explain its raison d'être, please?

Reply:
Fixed in https://github.com/apache/logging-log4j2/pull/2836/commits/d2bdd02ca9add9cc155d7c8a65f1335bf8d5dd70.

The `FixedStringLayout.java.source` file is also used by `ResolverUtilTest` to compile multiple versions of the same class.